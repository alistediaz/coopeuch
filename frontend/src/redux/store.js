import { configureStore } from "@reduxjs/toolkit";
import { reducer } from "./tareaSlice";

const store = configureStore({
  reducer: reducer,
});

export default store;