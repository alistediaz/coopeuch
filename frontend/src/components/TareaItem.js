import React, { useRef, useState } from "react";
import { deleteTarea, updateTarea } from "../redux/tareaSlice";
import { selectTareaById } from "../redux/tareaSlice";
import { useDispatch, useSelector } from "react-redux";
import {
  AiOutlineForm,
  AiFillSave,
  AiOutlineCarryOut,
  AiOutlineDelete,
  AiOutlineStop,
} from "react-icons/ai";

export const TareaItem = (props) => {
  const { item } = props;
  const [isEditing, setIsEditing] = useState(false);
  const inputRef = useRef(true);
  const tarea = useSelector((state) => selectTareaById(state, item.id));
  const dispatch = useDispatch();

  const changeFocus = () => {
    inputRef.current.disabled = false;
    inputRef.current.focus();
    setIsEditing(!isEditing);
  };

  const callUpdateTarea = async (tarea) => {
    await dispatch(updateTarea(tarea));
  };

  const update = (id, value) => {
    if (value === "") {
      alert("Debe ingresar descripciÃ³n de la tarea");
      inputRef.current.value = tarea.descripcion;
    } else {
      if (value !== tarea.descripcions) {
        callUpdateTarea({ id, descripcion: value, vigente: tarea.vigente });
      }
      setIsEditing(false);
      inputRef.current.disabled = true;
    }
  };

  const updateVigente = (item) => {
    callUpdateTarea({ ...item, vigente: !item.vigente });
  };

  const eliminaTarea = async (tareaId) => {
    await dispatch(deleteTarea(tareaId));
  };
  return (
    <li
      key={item.id}
      className="card"
      style={!item.vigente ? { backgroundColor: "Ivory" } : {}}
    >
      <textarea
        ref={inputRef}
        disabled={inputRef}
        defaultValue={item.descripcion}
        maxLength={100}
      />
      <div className="btns">
        {!isEditing && (
          <button onClick={() => changeFocus()} title="Editar">
            <AiOutlineForm />
          </button>
        )}
        {isEditing && (
          <button
            onClick={() => update(item.id, inputRef.current.value)}
            title="Grabar"
          >
            <AiFillSave />
          </button>
        )}
        {item.vigente && (
          <button onClick={() => updateVigente(item)} title="Vigente">
            <AiOutlineCarryOut />
          </button>
        )}
        {!item.vigente && (
          <button
            style={{ color: "grey" }}
            onClick={() => updateVigente(item)}
            title="No Vigente"
          >
            <AiOutlineStop />
          </button>
        )}

        <button style={{ color: "red" }} onClick={() => eliminaTarea(item.id)} title="Eliminar">
          <AiOutlineDelete />
        </button>
      </div>

      {item.completed && <span className="completed">done</span>}
    </li>
  );
};
