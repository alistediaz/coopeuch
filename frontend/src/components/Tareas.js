import React, { useState, useEffect } from "react";
import { addNewTarea } from "../redux/tareaSlice";
import { useDispatch, useSelector } from "react-redux";

export const Tareas = () => {
  const [tarea, setTarea] = useState("")
  const [vigente, setVigente] = useState(false)
  const [disabledButton, setDisabledButton] = useState(false)
  // @ts-ignore
  const tareaStatus = useSelector((state) => state.status)
  const dispatch = useDispatch()

  const add = async () => {
    if (tarea === "") {
      alert("Debe ingresar descripciÃ³n de la tarea");
    } else {
      setDisabledButton(true)
      await dispatch(
        addNewTarea({
          descripcion: tarea,
          vigente: vigente,
        })
      ).unwrap();
      setTarea("");
      setVigente(false);
    }
  };

  useEffect(() => {
    if (tareaStatus === "idle" || tareaStatus === "succeeded") {
      setDisabledButton(false)
    }
  }, [tareaStatus, dispatch]);

  const handleChange = (e) => {
    setTarea(e.target.value);
  };

  const handleChangeVigente = (e) => {
    setVigente(e.target.checked);
  };

  return (
    <div className="addTareas">
      <input
        type="text"
        onChange={(e) => handleChange(e)}
        className="tarea-input"
        value={tarea}
        maxLength={100}
        title="DescripciÃ³n"
      />
      <div className="vigente">
        Vigente:
        <input type="checkbox" onChange={(e) => handleChangeVigente(e)} />
      </div>
      <button className="add-btn" onClick={() => add()} disabled={disabledButton}>
        Nueva
      </button>
    </div>
  );
};
