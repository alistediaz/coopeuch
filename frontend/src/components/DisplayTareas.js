import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { selectAllTareas, fetchTareas } from "../redux/tareaSlice";
import { TareaItem } from "./TareaItem";

export const DisplayTareas = () => {
  const dispatch = useDispatch();
  // @ts-ignore
  const tareaStatus = useSelector((state) => state.status);
  // @ts-ignore
  const error = useSelector((state) => state.error);
  const tareas = useSelector(selectAllTareas);

  useEffect(() => {
    if (tareaStatus === "idle") {
      dispatch(fetchTareas());
    }
  }, [tareaStatus, dispatch]);

  let content;

  if (tareaStatus === "loading") {
    content = <h3>Leyendo...</h3>;
  } else if (tareaStatus === "succeeded") {
    content = tareas.map((tarea) => {
      return <TareaItem key={tarea.id} item={tarea} />;
    });
  }

  return (
    <div className="displaytareas">
      {tareaStatus === "failed" && <div>{error}</div>}
      <ul>{content}</ul>
    </div>
  );
};
