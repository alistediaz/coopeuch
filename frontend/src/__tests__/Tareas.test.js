import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import store from '../redux/store'
import { Tareas } from '../components/Tareas'

describe('<Tareas />', () => {
  const defaultProps = {}
  const wrapper = renderer.create(
    <Provider store={store}>
     <Tareas {...defaultProps} />
    </Provider>,
  )

  test('render', () => {
    expect(wrapper).toMatchSnapshot()
  })
})