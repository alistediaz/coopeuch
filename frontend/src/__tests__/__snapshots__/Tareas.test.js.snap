// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tareas /> render 1`] = `
<div
  className="addTareas"
>
  <input
    className="tarea-input"
    maxLength={100}
    onChange={[Function]}
    title="DescripciÃ³n"
    type="text"
    value=""
  />
  <div
    className="vigente"
  >
    Vigente:
    <input
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <button
    className="add-btn"
    disabled={false}
    onClick={[Function]}
  >
    Nueva
  </button>
</div>
`;
